docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,272 @@
+
+When K rules are translated to Kore axioms, the `?` variables are existentially
+quantified at the top of the RHS and the `ensures` clause is and-ed to the RHS.
+
+This design document presents four typical language constructs or mathematical
+functions that have unspecified or nondeterministic behaviors and shows their
+axiomatization in Kore.
+
+We also propose the corresponding K frontend syntax to define some of these
+behaviors.
+
+### Implementation concerns
+
+Except for the `fresh` variants (Example B), none of the other are currently
+supported by either the K frontend of the Haskell backend.
+
+
+## Example A: Random Number Construct `rand()`
+
+The random number construct `rand()` is a language construct which could be
+easily conceived to be part of the syntax of a programming language:
+
+```
+Exp ::= "rand" "(" ")"
+```
+
+### Axioms
+
+The intended semantics of `rand()` is that it can rewrite to any integer in
+a single step.
+Here is its axiom
+
+```
+(RAND) \forall I:Int . rand() -> * I
+// equivalent form(s)
+//     rand() -> \forall I:Int . * I
+//     rand() -> * I
+// non-equivalent form(s)
+//     rand() -> * (\forall I:Int . I)
+```
+
+Note that the `(RAND)` axiom is logically equivalent to the following infinitely
+many axioms.
+
+```
+(RAND0)  rand() -> * 0
+(RAND1)  rand() -> * 1
+(RAND2)  rand() -> * 2
+  ...    ...
+(RAND-1) rand() -> * (-1)
+(RAND-2) rand() -> * (-2)
+  ...    ...
+```
+
+One can define variants of `rand()` by further constraining the output variable
+as a precondition to the rule.
+For example, `randPrime()`, defined below, can rewrite to any _prime number_.
+
+```
+(RANDPRIME) \forall P:Int . isPrime(P) -> (randPrime() -> * P)
+// equivalent form(s)
+//          \forall P:Int . (randPrime() /\ isPrime(P)) -> *P
+//          \forall P:Int . randPrime() -> (isPrime(P) -> * P)
+//          randPrime() -> \forall P:Int . (isPrime(P) -> * P)
+//          remove \forall P:Int in all above three
+```
+
+where `isPrime(_)` is a predicate that can be defined in the usual way.
+
+As another example, `randDifferentFromList(Is)` takes a list `Is` of integers
+and can rewrite in one step to any integer that is not in `Is`.
+Here is its axiom
+
+```
+(RANDLIST) \forall Is:List{Int} . \forall I:Int . \neg(I \in Is) -> randDifferentFromList(Is) -> * I
+```
+
+### Frontend Syntax
+
+Since all variables are universally quantified, `rand()` and its variants could
+be written in the K frontend using regular rules and variables:
+
+```
+syntax Exp ::= rand ()
+rule rand() => X:Int
+
+syntax Exp ::= randPrime ()
+rule randPrime() => X:Int
+  requires isPrime(X)
+
+syntax Exp ::= randList (List{Int})
+rule randList(Is) => I:Int
+  requires notBool (I inList{Int} Is)
+```
+
+Note 1: all above are not function symbols, but languages constructs.
+
+Note 2: Currently the frontend does not allow rules with universally quantified
+variables in the RHS which are not bound in the LHS.
+
```

Comment:
The above looks good to me.  So basically you would use a fresh, non-? variable in the RHS. The frontend may still issue a warning, just in case the user meant the other case below.

Reply:
All these rules are non-executable in a concrete setting, therefore, we should maybe add attributes to them to signal that. Moreover, it might be useful if we use different attribute names because this would help the compiler better identify the rules and process them accordingly.